gallery/experiments/simulated_abinitio_pipeline.py

```diff
@@ -0,0 +1,241 @@
+# used in the remainder of the reconstruction pipeline.
+#
+# In this example, this behavior is controlled by the `do_cov2d` boolean variable.
+# When disabled, the original src and default aligner is used.
+# If you will not be using cov2d,
+# you may remove this code block and associated variables.
+
+classification_src = src
+custom_averager = None
+if do_cov2d:
+    # Use CWF denoising
+    cwf_denoiser = DenoiserCov2D(src)
+    # Use denoised src for classification
+    classification_src = cwf_denoiser.denoise()
+    # Peek, what do the denoised images look like...
+    if interactive:
+        classification_src.images(0, 10).show()
+
+    # Use regular `src` for the alignment and composition (averaging).
+    composite_basis = FFBBasis2D((src.L,) * 2, dtype=src.dtype)
+    custom_averager = BFSReddyChatterjiAverager2D(composite_basis, src, dtype=src.dtype)
+
+
+# %%
+# Class Averaging
+# ----------------------
+#
+# Now perform classification and averaging for each class.
+
+logger.info("Begin Class Averaging")
+
+rir = RIRClass2D(
+    classification_src,  # Source used for classification
+    fspca_components=400,
+    bispectrum_components=300,  # Compressed Features after last PCA stage.
+    n_nbor=n_nbor,
+    n_classes=n_classes,
+    large_pca_implementation="legacy",
+    nn_implementation="sklearn",
+    bispectrum_implementation="legacy",
+    averager=custom_averager,
+)
+
+classes, reflections, distances = rir.classify()
+avgs = rir.averages(classes, reflections, distances)
+if interactive:
+    avgs.images(0, 10).show()
+
+# %%
+# Common Line Estimation
+# ----------------------
+#
+# Next create a CL instance for estimating orientation of projections
+# using the Common Line with Synchronization Voting method.
+
+logger.info("Begin Orientation Estimation")
+
+# Stash true rotations for later comparison,
+#   note this line only works with naive class selection...
+true_rotations = src.rots[:n_classes]
+
+orient_est = CLSyncVoting(avgs, n_theta=36)
+# Get the estimated rotations
+orient_est.estimate_rotations()
+rots_est = orient_est.rotations
+
+logger.info("Compare with known rotations")
+# Compare with known true rotations
+Q_mat, flag = register_rotations(rots_est, true_rotations)
+regrot = get_aligned_rotations(rots_est, Q_mat, flag)
+mse_reg = get_rots_mse(regrot, true_rotations)
+logger.info(
+    f"MSE deviation of the estimated rotations using register_rotations : {mse_reg}\n"
+)
+
+# %%
+# Volume Reconstruction
+# ----------------------
+#
+# Using the estimated rotations, attempt to reconstruct a volume.
+
+logger.info("Begin Volume reconstruction")
+
+# Assign the estimated rotations to the class averages
+avgs.rots = rots_est
+
+# Create a reasonable Basis for the 3d Volume
+basis = FFBBasis3D((v.resolution,) * 3, dtype=v.dtype)
+
+# Setup an estimator to perform the back projection.
+estimator = MeanEstimator(avgs, basis)
+
+# Perform the estimation and save the volume.
+estimated_volume = estimator.estimate()
```

Comment:
I can try it. I'll come back to this after the other items.  I didn't have any idea that `Q_mat` would realize the right rotations for that...  I'm not sure the project is there yet (good rmse), lots of things to check, but I was reasonably happy to get things like the attached at this stage of the project. First is the original hi res map from EMD 10028, and second is the simulated abinitio model, both as viewed in chimera.

<img width="759" alt="Screen Shot 2022-02-18 at 7 04 18 AM" src="https://user-images.githubusercontent.com/47759732/154680571-b06c14ca-800a-4d79-bab1-893bd39211bd.png">
<img width="814" alt="Screen Shot 2022-02-18 at 7 09 49 AM" src="https://user-images.githubusercontent.com/47759732/154680575-6cbcfba2-b72d-4552-9b93-c49c19a21794.png">

Reply:
FWIW that sim was `gallery/experiments/simulated_abinitio_pipeline.py`, if you'd like to run it too. Just requires downloading the map file and putting in the directory.

Reply:
That doesn't look too bad, actually…

So the idea is that `Q_mat` tells us how the estimated rotations differ from the ground truth rotations (in other words, what rotation needs to be applied to transform one rotation sequence to the other). Thing that might be tricky is that it may include a possible reflection as well. It's definitely a “nice to have” and not a requirement. If it takes more than fifteen minutes, make an issue and we'll come back to it.

Reply:
Yes, I was quite pleased with it for the maturity of the project and sort of rushing this last feature given the other situations.  Lot of code refinement todo, but I think this should get the project down the road quite a ways.

I'll load the saved volume and give it a try.